Core/Modules/FSM/HFSM_State.cs

```diff
@@ -0,0 +1,57 @@
+namespace Squiggles.Core.FSM;
+
+using Godot;
+
+/// <summary>
+/// Serves both the role of <see cref="FiniteStateMachine"/> and <see cref="State"/>. This can be used to create a naive Hierarchical State Machine, which allows nesting smaller state machines within a larger state machine in order to produce more complex behaviour without requiring a mess of transition definitions.
+/// </summary>
+/// <remarks>
+/// I really dislike this name. Anything with an underscore makes me cringe (at least for C# type names). I'm tempted to just write it all out so it's more obvious and removes the underscore.
+/// </remarks>
+[GlobalClass]
+public partial class HFSM_State : State {
+
+  [Export] private State _initialState;
+
+  private State _current;
+
+  public override void _Ready() {
+    // reduce overhead
+    SetProcess(false);
+    SetPhysicsProcess(false);
+
+    // load inital state
+    ChangeState(_initialState);
+  }
+
+  public virtual void ChangeState(State n_state) {
+    if (_current is not null) {
+      _current.IsActive = false;
+      _current.ExitState();
+      _current.OnStateFinished -= HandleChildStateFinished;
+    }
+    _current = n_state;
+    if (_current is not null) {
+      _current.EnterState();
+      _current.IsActive = IsActive; // only set active if this node is also active. Prevents inactive nodes performing processing
+      _current.OnStateFinished += HandleChildStateFinished;
+    }
+  }
+
+  public override void EnterState() {
+    if (_current is null) {
+      ChangeState(_initialState);
+    }
+    _current?.EnterState();
```

Comment:
I have seen a couple places where you have this pattern of `_something?.DoIt();` followed by `_something.Blah = bloop`

The null conditional operator `?.` basically adds a null check, and if the variable is null, it doesn't explode while calling the method, it just skips it. It transforms in to this:
```
if (_current != null)
{ 
    _current.EnterState(); 
}
```
Which means that you think it could be null at this point. This means that by then accessing `_current` right after it you are then saying it can't be null. So in general you should have all accesses to a variable either do `.` or `?.` based on any null checks you have and what you think it should be.

Reply:
I suppose for that case, I am assuming the possibility that `_initialState` is null (because the dev forgot to set it in the Godot editor). Though perhaps causing an error would be more helpful and intuitive.